### DISCUSSION TITLE
Historical Playback, Benchmarking, InfluxDB Questions

### DISCUSSION THREAD

**Discussion by altsv** [nasa/openmct#7445] [Community Q&A] [answered]
Hello!

I am part of a software team that has been tasked with developing a visualization application for monitoring telemetry data from multiple sources. We have diligently worked through all of the excellent tutorials and resources you’ve provided to determine that we’d like to proceed forward with using OpenMCT as a real-time data visualization tool. Having said that, we find ourselves with a few questions for our use case:

1. Does Open MCT support historical playback? If so, how would we enable or implement this? We're looking to play through historical data similarly to the Realtime mode.

2. Would you happen to have any releasable documentation on performance metrics and/or test plans used to derive any performance data? To ensure we understand any performance limitations, we’re transitioning to a benchmarking phase for several different production environment scenarios. We were able to find a Usability Study online conducted by Blue Origin in 2021, but the report is UI/UX centric and doesn’t cover any end-to-end performance metrics.  If any reports carry any releasability caveats, hopefully we can work through applicable channels to obtain those as well. Anything you could provide would be greatly appreciated.
 
3. Do you have any information on integrating Open MCT with InfluxDB? We noticed in the openmct-performance repo that there is some commented code for taking in data from an InfluxDB. 

Thank you so much,
Simon

**Comment by ozyx** [comment, marked as answer]
Hi Simon,

Great to hear that your team is considering using Open MCT! Please find our responses below-- I'll also tag the POCs who can best answer any follow-ups for those specific questions.


> 1. Does Open MCT support historical playback? If so, how would we enable or implement this? We're looking to play through historical data similarly to the Realtime mode.

@akhenry:
The answer to this question kind of depends on what you're using to feed the data into Open MCT. Open MCT is primarily a visualization tool, so it will display whatever data is fed into it. As long as your backend can play back historical data, Open MCT can display it. We do support a "remote clock" which can be configured to tick on a remote time source so that a playback clock can be used to drive the time conductor in real-time mode.

> 2. Would you happen to have any releasable documentation on performance metrics and/or test plans used to derive any performance data? To ensure we understand any performance limitations, we’re transitioning to a benchmarking phase for several different production environment scenarios. We were able to find a Usability Study online conducted by Blue Origin in 2021, but the report is UI/UX centric and doesn’t cover any end-to-end performance metrics. If any reports carry any releasability caveats, hopefully we can work through applicable channels to obtain those as well. Anything you could provide would be greatly appreciated.

@unlikelyzero:
We do not have any externally available performance test reports for comparing one telemetry source to another, nor have we encountered benchmarks for comparing open-source mission control software. I am interested in understanding your approach to this benchmarking.

Consider the following factors:

1. The number of tabs a user will have open
1. The hours before a refresh is needed
1. The total number of users
1. Network bandwidth constraints per client and per deployment
1. Web server hosting
1. Real-time versus historical data, and query sizes (15 minutes versus 15 days)

Just as we offer a platform for building automated acceptance tests for your implementation of Open MCT ([openmct-as-a-dependency](https://github.com/unlikelyzero/openmct-as-a-dependency)), we also provide the [openmct-performance](https://github.com/unlikelyzero/openmct-performance.git) repository as a universal platform for testing Open MCT performance in the context of your environment. Open MCT has been around for a long time, and when performance issues are reported, we can support the necessary changes. Often, Open MCT becomes the bearer of bad news regarding general performance issues, as it serves as the client interface. We maintain a suite of [regression performance tests](https://github.com/nasa/openmct/tree/master/e2e/tests/performance) designed to ensure that our application's client performance does not degrade as we introduce new features. Additionally, this suite helps us detect Chrome Client performance regressions.

> 3. Do you have any information on integrating Open MCT with InfluxDB? We noticed in the openmct-performance repo that there is some commented code for taking in data from an InfluxDB.

@unlikelyzero:
We currently do not have a working example for integrating Open MCT with InfluxDB. However, InfluxDB offers a simple Node.js client that could likely facilitate a straightforward integration. This integration was initially considered and even commented out of the performance repository because we planned to switch the [openmct-performance](https://github.com/unlikelyzero/openmct-performance) repository from Prometheus to InfluxDB. However, our plans changed when Prometheus implemented the remote_write interface. For assistance in writing a backend adapter for Open MCT, we offer a tutorial [here](https://github.com/nasa/openmct-tutorial?tab=readme-ov-file#part-c---integrateproviderequest-telemetry). Please note the tutorial is due for an update, but the fundamentals are still relevant.

---

Hopefully that demystifies things a bit. Please feel free to comment in this thread with any follow-up questions and we will do our best to answer them in a timely manner.

Cheers!

Jesse

**Reply by serbuh** [reply]
@ozyx , Do you have any example of substituting LocalClock with the RemoteClock?

**Reply by ozyx** [reply]
Sure. So the RemoteClock really just requires something that ticks. You would need to provide a telemetry object that subscribes to some data source providing a ticking value (like a timestamp or ms since epoch for example). You'll also need telemetry providers for both historical and real-time data that support `request` and `subscribe` for this object type.

From there, you would install the `RemoteClock()` plugin, specifying the Identifier of the telemetry object that provides your tick:

```js
openmct.install(openmct.plugins.RemoteClock({
    namespace: 'myTelemetry',
    key: 'myRemoteClock'
}));
```

Then you just need to register the clock with the TimeConductor:
```js
        openmct.install(openmct.plugins.Conductor({
            menuOptions: [
                {
                    name: "Remote",
                    timeSystem: 'utc',
                    clock: 'remote-clock',
                    clockOffsets: {
                        start: - FIFTEEN_MINUTES,
                        end: FIFTEEN_SECONDS
                    }
                },
                // your other clocks here
            ]
```

Hope this helps!

**Reply by serbuh** [reply]
Thank you, @ozyx . Your examples always help!
Started to implement external tics.
I have added what you mentioned. And also added two plugins (realtime and telemetry) for the new telemetry type. 
Currently getting the following error:
```
Uncaught Error: No Provider Matched for keyString "myTelemetry:myRemoteClock"
    at ObjectAPI.get (ObjectAPI.js:241:1)
    at RemoteClock.start (RemoteClock.js:62:1)
    at RemoteClock.on (DefaultClock.js:59:1)
    at TimeAPI.setClock (TimeContext.js:544:1)
    at eval (plugin.js:131:1)
    at MCT.install (MCT.js:368:1)
    at (index):134:15
```
Can you point out what I am missing here?

Thank you

**Reply by ozyx** [reply]
> Can you point out what I am missing here?

Yep! So you'll just need to replace "myTelemetry" and "myRemoteClock" with the actual namespace and key for your remote tick source as defined by your object provider.

**Reply by alexhoffenreich** [reply]
Hi @ozyx ,
Just for clarification,
Does it mean that the remote clock realtime and historic providers have to handle separate listeners (subscribe/unsubscribe) and when the telemetry point arrives from the server they should also call the callback with point (just like the regular realtime and historic providers)?
If so, the remote clock providers and the regular providers are almost identical.
Correct me if i am wrong:
Regular provider receives a telemetry point from the server, sends the point to a plot / table / gauge  that displays the data based on the current client clock and the conductor.
Remote clock providers receives a telemetry point from the server, sends the point to a plot / table / gauge  that displays the data based on the timestamp in the point (like a fake real-time clock) and the conductor.

**Reply by serbuh** [reply]
@ozyx Easier said than done 😅 

Here is my code:
#### index.html
```js
...
// Remote clock plugins install
openmct.install(openmct.plugins.RemoteClockHistoricalTelemetryPlugin('RemoteClockDomainObject'));
openmct.install(openmct.plugins.RemoteClockRealtimeTelemetryPlugin('RemoteClockDomainObject'));
openmct.install(openmct.plugins.RemoteClockProviderPlugin());

openmct.install(openmct.plugins.RemoteClock({
    namespace: 'RemoteClockMainspace',
    key: 'RemoteClockKey'
}));

openmct.install(
  openmct.plugins.Conductor({
    menuOptions: [
      {
        name: "Remote",
        timeSystem: 'utc',
        clock: 'remote-clock',
        clockOffsets: {
            start: - FIFTEEN_MINUTES,
            end: THIRTY_SECONDS
        }
      },
      // other clocks...
    ]
  })
)
...
```
#### Remote Clock Providers plugin
```js
let object_name = 'PredefinedTelemetry';
let object_type = 'RemoteClockDomainObject';
let object_namespace = 'RemoteClockMainspace';

export default function RemoteClockProviderPlugin() {
  return function install(openmct) {
    // An object provider builds Domain Objects
    openmct.objects.addProvider(object_namespace, {
      get: function (identifier) {
        console.log('provider.get( ' + identifier.key + ' )');
        return Promise.resolve({
          identifier: identifier,
          name: object_name,
          type: object_type
        });
      }
    });

    // Composition provider
    openmct.composition.addProvider({
      appliesTo: function (domainObject) {
        return false;
      },
      load: function (domainObject) {
        return null;
      }
    });

    // Add RemoteClock type
    console.log('AddType ' + object_type);

    openmct.types.addType(object_type, {
      name: 'RemoteClockTick',
      description: 'RemoteClockTick',
      cssClass: 'icon-telemetry'
    });
  };
}
```

#### RemoteClock Realtime Plugin

```js
let object_key = 'RemoteClockKey';
export default function RemoteClockRealtimeTelemetryPlugin(desired_domain_object_type) {
  return function (openmct) {
    let listeners = {};
    let socket = io();
    my_socket.on('realtime-tick', (msg) => {
      // Get realtime message
      console.log('Tick: ', msg.timestamp); /// This really happens every second

      /// TODO How exactly can we inform API about new tick?

    });

    let provider = {
      supportsSubscribe: function (domainObject) {
        console.log('supportsSubscribe ', domainObject.identifier.key);
        return domainObject.type === desired_domain_object_type;
      },
      subscribe: function (domainObject, callback, options) {
        console.log('Subscribe to RemoteClock', domainObject.identifier.key);

        /// This never happens :/

      }
    };

    openmct.telemetry.addProvider(provider);
  };
}
```

#### RemoteClock Historical plugin
```js
export default function RemoteClockHistoricalTelemetryPlugin(desired_domain_object_type) {
  return function install(openmct) {
    let provider = {
      supportsRequest: function (domainObject) {
        return domainObject.type === desired_domain_object_type;
      },
      request: async function (domainObject, options) {
        let url = \`/history/${domainObject.identifier.key}/${options.start}/${options.end}/${options.strategy}/${options.size}\`;
        const resp = await fetch(url);
        const data = await resp.json();
        console.log('Got ' + data.length + ' history items for ' + domainObject.identifier.key);
        return data;
      }
    };
    openmct.telemetry.addProvider(provider);
  };
}
```

What do I get:
```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'value')
    at RemoteClock._timeSystemChange (RemoteClock.js:139:1)
    at eval (RemoteClock.js:73:1)
```
The problem that `this.metadata` was never defined in line:
```
let metadataValue = this.metadata.value(timeKey);
```
As a result `this._subscribe();` of the RemoteClock never happens..
Would be grateful for any help.
Thanks!
